File: 2779-number-of-adjacent-elements-with-the-same-color/solution.ts
function colorTheArray(n: number, queries: number[][]): number[] {
  const arr = Array.from({ length: n }).map(() => 0);
  return queries.reduce((acc, cur) => {
    const [idx, color] = cur;
    let sameColor = acc.slice(-1)[0] || 0;
    if (arr[idx - 1] === arr[idx] && arr[idx] !== 0) sameColor--;
    if (arr[idx] === arr[idx + 1] && arr[idx] !== 0) sameColor--;
    arr[idx] = color;
    if (arr[idx - 1] === arr[idx]) sameColor++;
    if (arr[idx] === arr[idx + 1]) sameColor++;
    acc.push(sameColor);
    return acc;
  }, []);
}

File: 2779-number-of-adjacent-elements-with-the-same-color/solution.cpp
class Solution {
public:
    vector<int> colorTheArray(int n, vector<vector<int>>& queries) {
        int now = 0;
        vector<int> ans, colors(n, 0);
        for(auto& q : queries) {
            int idx = q[0], color = q[1];
            if(colors[idx] != color) {
                if(idx > 0 && colors[idx - 1] > 0 && colors[idx - 1] == colors[idx]) now--;
                if(idx < n - 1 && colors[idx + 1] > 0 && colors[idx] == colors[idx + 1]) now--;
                if(idx > 0 && colors[idx - 1] == color) now++;
                if(idx < n - 1 && color == colors[idx + 1]) now++;
            }
            ans.push_back(now);
            colors[idx] = color;
        }
        return ans;
    }
};
